### ISSUE TITLE
Complex GET query param objects in Swashbuckle

### ISSUE THREAD

**Issue by pootzko** [pootzko/tkit.dev#8]
https://tkit.dev/2020/03/10/complex-get-query-param-objects-in-swashbuckle/

**Comment by mmalolepszy** [comment]
Nice one! I just started using this filter and it looks like it does the job.

One issue I noticed is when you use `CancellationToken` as a controller parameter. For example:
```
public async Task<ActionResult<SomeResult>> Get([FromQuery] SomeRequest request, CancellationToken cancellationToken)
```
In that case `CancellationToken` is also generated as a model. Swashbuckle removes it by default (https://github.com/domaindrivendev/Swashbuckle.WebApi/commit/7ab9b842c5ad068dd7115a102bed45559133e0a6) but your filter doesn't. It can be fixed by adding condition to action parameters enumeration:
```
var actionParameters = description.ActionDescriptor.Parameters
    .Where(p => p.ParameterType != typeof(CancellationToken))
    .ToList();
```

**Comment by tihomir-kit** [comment]
Oh nice, thanks for pointing that out. I'll update the blog post and point it to your comment if someone else needs this as well. Thanks.

**Comment by ksydex** [comment]
@tihomir-kit Hi! Thank you for post but it seems like it doesn't working in my case.
When i register this filter and use it:
![image](https://user-images.githubusercontent.com/38377534/118737807-4d5f8e80-b84e-11eb-8d6f-b42033ecf324.png)
Binding like this:
![image](https://user-images.githubusercontent.com/38377534/118737841-5cded780-b84e-11eb-880e-05191fbe56bd.png)
The interface itself is:
![image](https://user-images.githubusercontent.com/38377534/118737912-813ab400-b84e-11eb-8a79-9afe873c5176.png)
Any reason why it's not working?
Thanks in advance
P.S. Using net 5.0.1

**Comment by tihomir-kit** [comment]
Hi, what's not working exactly? 

The only glaring difference compared to how I use it is that you used an interface and I use a class..

**Comment by ksydex** [comment]
@tihomir-kit well, it just doesn't show parms as it supposed to be. 
I'll try to use class instead and let you know if it will work

**Comment by ksydex** [comment]
I tried to use a class and still the same.
The class is
![image](https://user-images.githubusercontent.com/38377534/118894534-19976e00-b90d-11eb-98ac-fe4a817c34b3.png)
And in swagger ui it's like that 
![image](https://user-images.githubusercontent.com/38377534/118894566-26b45d00-b90d-11eb-8aba-26252feffd7c.png)

**Comment by tihomir-kit** [comment]
"it just doesn't show parms as it supposed to be" 

What would you expect to see, and where? What are you seeing instead of what you're expecting to see? Have you tried making a request against the API? Did it work, did it fail, did params get read, and could you access them inside the controller?

**Comment by ksydex** [comment]
@tihomir-kit i expect to see documentated params from the class. I.e:
![image](https://user-images.githubusercontent.com/38377534/119271868-1e5d6a00-bc0c-11eb-9c93-e5147c2cd755.png)
(these params on the screenshot are from inline c# method params)
Instead i get this:
![image](https://user-images.githubusercontent.com/38377534/119271845-0b4a9a00-bc0c-11eb-9307-48ed6360d12c.png)

**Comment by tihomir-kit** [comment]
Oh, ok, I see. I'm not sure to be honest with you. You might have to tweak the middleware integration a bit. But I would expect the requests (when consumed from within JS/TS) to still work just fine. This is just how Swagger displays it on the documentation page. I personally don't worry about these kinds of details as long as it's functional, but that's just me.

**Comment by ksydex** [comment]
@tihomir-kit yeah, requests will work of course. I'll try to tweak the code. Thanks

**Comment by mbarmettler** [comment]
thanks @tihomir-kit 
was able to get my QueryParam Models generated with your code - saved me alot of time and research, since its not very well documented (or at least i could not find practicable examples).
what surprises me furthermore, is that QueryPram Models with inheritance, are not generated like others with enhitance.
The Properties are getting flattend anyway on thes Query Pram Models - on others not (like i used the "UseAllOfForInheritance();" option.
how to deal with this? does the order matter in startup.cs?

**Comment by tihomir-kit** [comment]
@mbarmettler Glad to help. 

Flattened? Are we talking about inheritance or nested objects inside your root query params object? Inheritance works fine for me, no additional setup (compared to blog post) required. Nested sub-objects, I'm not sure, I try to keep my query params flat to avoid extra complexity and problems. So not sure really, sorry!

**Comment by mbarmettler** [comment]
@tihomir-kit 
thanks for your help on this.
what i mean is:
we have QueryModels with inheritance which is not generated correctly

The Model:
`TimelineHolidayQueryModel : TimeLineQuery`

what i get generated:
`export interface TimelineHolidayQueryModel {
    pageNumber: number;
    pageSize: number;
    filterFullText: boolean;
    full: boolean;
    sortBy: string | undefined;
    sortOrder: string | undefined;
    authorisation: boolean;
    includeNullPerson: boolean;
    authorizedPerson: number[] | undefined;
    pageDistinct: boolean;
    pageDistinctField: string | undefined;
    personID: number | undefined;
    id: number | undefined;
    idFrom: number | undefined;
    idTo: number | undefined;
`

all fields to "ID" are those from the TimeLineQuery - which gets not generated even though i have configured like this on startup.cs:
`//adding addional Swashbuckle / OpneApi configurations
                c.SchemaFilter<NSwagEnumExtensionSchemaFilter>();
                c.OperationFilter<FromQueryModelFilter>();
                c.UseAllOfForInheritance();
                c.UseOneOfForPolymorphism();
                c.SelectSubTypesUsing(baseType =>
                {
                    return typeof(Startup).Assembly.GetTypes().Where(type => type.IsSubclassOf(baseType));
                });`

and also this:
`app.UseEndpoints(endpoints => {
    endpoints.MapControllers();
    endpoints.MapSwagger();
    endpoints.MapControllerRoute(
        name: "default",
        pattern: "Report/{controller=Home}/{action=Index}/{id?}");
}`

i was expecting proper inheritance generation with setting "UseAllOfForInheritance" - but not.
does the order by the way matter in the config section?

Thanks for any help or hints

**Comment by tihomir-kit** [comment]
No idea TBH, sorry.
